Fix memory leak in strbuf.

aronsky opened the pull request:
`sb->ptr` was set to NULL without freeing the underlying data.

ret2libc commented on libr/anal/fcn.c:
```diff
@@ -2301,13 +2301,13 @@ R_API ut32 r_anal_fcn_cost(RAnal *anal, RAnalFunction *fcn) {
 		int idx = 0;
 		for (at = bb->addr; at < end;) {
-			memset (&op, 0, sizeof (op));
+			// memset (&op, 0, sizeof (op)); // Not needed, op = R_EMPTY
```
So, IMHO this shouldn't be needed at all ever, because `r_anal_op` should set all the necessary fields. However, setting `op = R_EMPTY` outside the loop is not the same as zeroing it at each iteration.

That said, if you want to remove the line, just remove it instead of uncommenting please.

aronsky replied:
You're absolutely right, I did a search for all the `RAnalOp`s defined on the stack, and removed the following `memset`s where appropriate - didn't notice this one is in a loop. I will restore it, as well as remove the other `memset`s instead of commenting them out.

ret2libc commented on libr/anal/op.c:
```diff
@@ -14,14 +14,7 @@ struct VarUsedType {
 
 R_API RAnalOp *r_anal_op_new () {
 	RAnalOp *op = R_NEW0 (RAnalOp);
```
use R_NEW since you are doing `memset (..0, ..)` in `r_anal_op_init`.

ret2libc commented on libr/anal/reflines.c:
```diff
@@ -80,15 +80,15 @@ R_API void r_anal_reflines_free (RAnalRefline *rl) {
 	ut64 opc = addr;
 
-	memset (&op, 0, sizeof (op));
+	// memset (&op, 0, sizeof (op)); // Not needed, op = R_EMPTY
```
remove line.

ret2libc commented on libr/asm/op.c:
```diff
@@ -3,7 +3,9 @@
 
 R_API RAsmOp *r_asm_op_new() {
-	return R_NEW0 (RAsmOp);
+	RAsmOp *op = R_NEW0 (RAsmOp);
```
use `R_NEW`, memset is already in the init.

aronsky commented:
> The memory leak fix in strbuf looks ok.
> 
> However, were all the other changes for r_anal_op_init and r_asm_op_init really necessary? In any case, if we define them, I think it makes more sense to always call them instead of relying on R_EMPTY. R_EMPTY indeed sets all the Op fields to 0, which may not be what r_asm_op_init or r_anal_op_init do (e.g. r_anal_op_init does set some fields to UT64_MAX... in this particular case it won't probably matter, but we never know in the future what happens, so I'd be more inclined to always call r_*_op_init instead of using R_EMPTY).

I wanted `r_anal_op_init` and `r_asm_op_init` to be more in sync and parallel to each other. I agree that it isn't directly related to the PR's subject, per se - but the PR deals with the initialization of those objects on the stack, so I think it's relevant.

I also agree that using `R_EMPTY` might not be the best choice when `r_*_op_init` is available for initialization. However, the `r_*_op_init` functions are currently pretty redundant, and in-place initialization is far more readable.

If you think it's best to initialize them with `r_*_op_init`, I'll change that. I agree that it's better for future-proofing the code, and if anything relies on the RAnalOp fields being initialized to UT64_MAX, that'll take care of it, too.

ret2libc commented:
I'm thinking about it and... maybe r_anal_op_init and r_asm_op_init shouldn't be required at all. After all, r_asm_op and r_anal_op should ensure that `op` is always initialized to some sane defaults, even if there are errors. I think currently it's not clear whether the `memset(op,0)` are still required, so maybe it would be better to not touch them at all.

@radare what do you think?

radare commented:
when i did that i was initializing first in the plugin side, but then some plugins was not initializing it, so it was causing UB, so i endup memsetting that in the r_anal_op function.. i think thats the right place for this to happen.

also i think that r_asm_op should be a subset of r_anal_op, because r_anal_op allows to define the level of granularity in detail, so adding asm/dis in there shuold be enough to "kill" the duplicity between asm and anal.

> On 5 Mar 2019, at 19:03, Riccardo Schirone <notifications@github.com> wrote:
> 
> I'm thinking about it and... maybe r_anal_op_init and r_asm_op_init shouldn't be required at all. After all, r_asm_op and r_anal_op should ensure that op is always initialized to some sane defaults, even if there are errors. I think currently it's not clear whether the memset(op,0) are still required, so maybe it would be better to not touch them at all.
> 
> @radare <https://github.com/radare> what do you think?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/13273#issuecomment-469787530>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA3-lrGxC6SbB7aUONtc2wf170lniOsDks5vTrGBgaJpZM4beUt_>.
>

radare commented on libr/anal/op.c:
```diff
@@ -14,14 +14,7 @@ struct VarUsedType {
-		op->val = UT64_MAX;
-		r_strbuf_init (&op->esil);
-	}
+	r_anal_op_init(op);
```
missing space before (

ret2libc commented:
Ok I agree. I think the initialization should be done in r_anal_op/r_asm_op.

radare commented:
Yep

> On 5 Mar 2019, at 20:34, Riccardo Schirone <notifications@github.com> wrote:
> 
> Ok I agree. I think the initialization should be done in r_anal_op/r_asm_op.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

ret2libc commented:
@aronsky sorry your PR became object of discussion, but the more you modify, the riskier it is :D

So, to wrap up: memset before using r_asm_disassemble/r_anal_op are not required and neither initializing op to R_EMPTY. Indeed, `r_asm_disassemble` and `r_anal_op` should work as a sort of initializer and be sure `op` always contain consistent data. If you look at `r_anal_op` code it actually already does the memset (though it would be better to move the memset before the `r_return_val_if_fail` line, and maybe use your r_anal_op_init instead), while `r_asm_disassemble` does not call anything (though it should call memset or, even better, your `r_asm_op_init`).

Does that sound ok?

aronsky commented:
> @aronsky sorry your PR became object of discussion, but the more you modify, the riskier it is :D

No problem, discussion is welcome!

I reverted most of the changes, and added a call to `r_asm_op_init`/`r_anal_op_init` in `r_asm_disassemble`/`r_anal_op`, as suggested.

I left the new `r_anal_op_init` function, as I think it's correct for it to exist :)

I also called the initialization functions prior to `r_return_val_if_fail`, as per your suggestion (they are guarded against NULL pointers), thus making sure the Op is returned in an initialized state even on any failure.

ret2libc commented:
Thanks a lot! This looks great now ;)
